File: src/openjarvis/security/file_policy.py
"""File sensitivity policy — block access to secrets, credentials, and keys."""

from __future__ import annotations

from pathlib import Path
from typing import Iterable, List, Union

DEFAULT_SENSITIVE_PATTERNS: frozenset[str] = frozenset(
    {
        ".env",
        ".env.*",
        "*.env",
        ".secret",
        "*.secrets",
        "credentials.*",
        "*.pem",
        "*.key",
        "*.p12",
        "*.pfx",
        "*.jks",
        "id_rsa",
        "id_ed25519",
        ".htpasswd",
        ".pgpass",
        ".netrc",
    }
)


def is_sensitive_file(path: Union[str, Path]) -> bool:
    """Return ``True`` if *path* matches a sensitive file pattern.

    Checks both the filename and the full name of the resolved path against
    ``DEFAULT_SENSITIVE_PATTERNS`` using :func:`fnmatch.fnmatch`. Resolving
    existing paths prevents a harmless-looking symlink from bypassing the
    policy for a sensitive target such as ``.env``.
    Uses the Rust implementation when available, falls back to Python.
    """
    path_obj = Path(path)
    try:
        checked_path = path_obj.resolve(strict=False)
    except (OSError, RuntimeError):
        # Keep the original path for missing paths and symlink loops. The
        # filesystem operation will report its own error if the path is not
        # usable, while ordinary new files retain the existing policy.
        checked_path = path_obj

    try:
        from openjarvis._rust_bridge import get_rust_module

        _rust = get_rust_module()
        return _rust.is_sensitive_file(str(checked_path))
    except ImportError:
        return _is_sensitive_file_py(str(checked_path))


def _is_sensitive_file_py(path_str: str) -> bool:
    """Pure-Python fallback for sensitive file detection."""
    import fnmatch

    p = Path(path_str)
    name = p.name
    for pattern in DEFAULT_SENSITIVE_PATTERNS:
        if fnmatch.fnmatch(name, pattern) or fnmatch.fnmatch(str(p), pattern):
            return True
    return False


def filter_sensitive_paths(paths: Iterable[Union[str, Path]]) -> List[Path]:
    """Return only non-sensitive paths from *paths*."""
    return [Path(p) for p in paths if not is_sensitive_file(p)]


__all__ = [
    "DEFAULT_SENSITIVE_PATTERNS",
    "filter_sensitive_paths",
    "is_sensitive_file",
]

File: rust/crates/openjarvis-security/src/file_policy.rs
//! File sensitivity policy — block access to secrets, credentials, and keys.

use once_cell::sync::Lazy;
use std::collections::HashSet;
use std::path::{Path, PathBuf};

static SENSITIVE_PATTERNS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    HashSet::from([
        ".env",
        ".secret",
        "id_rsa",
        "id_ed25519",
        ".htpasswd",
        ".pgpass",
        ".netrc",
    ])
});

static SENSITIVE_EXTENSIONS: Lazy<Vec<&'static str>> = Lazy::new(|| {
    vec![
        ".pem", ".key", ".p12", ".pfx", ".jks", ".secrets",
    ]
});

static SENSITIVE_PREFIXES: Lazy<Vec<&'static str>> = Lazy::new(|| {
    vec![".env.", "credentials."]
});

/// Resolve a path for policy checks without requiring the final target to exist.
///
/// `canonicalize` handles normal existing paths and symlink chains. When the
/// final target does not exist (for example, a write through `notes.txt ->
/// .env`), follow the link entries manually so the sensitive target name is
/// still checked.
fn path_for_policy(path: &Path) -> PathBuf {
    if let Ok(resolved) = std::fs::canonicalize(path) {
        return resolved;
    }

    let mut candidate = path.to_path_buf();
    for _ in 0..40 {
        let Ok(target) = std::fs::read_link(&candidate) else {
            break;
        };
        candidate = if target.is_absolute() {
            target
        } else {
            candidate
                .parent()
                .unwrap_or_else(|| Path::new(""))
                .join(target)
        };
    }
    candidate
}

/// Return `true` if path or its resolved target matches a sensitive pattern.
pub fn is_sensitive_file(path: &Path) -> bool {
    let checked_path = path_for_policy(path);
    let name = match checked_path.file_name().and_then(|n| n.to_str()) {
        Some(n) => n,
        None => return false,
    };

    if SENSITIVE_PATTERNS.contains(name) {
        return true;
    }

    for ext in SENSITIVE_EXTENSIONS.iter() {
        if name.ends_with(ext) {
            return true;
        }
    }

    for prefix in SENSITIVE_PREFIXES.iter() {
        if name.starts_with(prefix) {
            return true;
        }
    }

    false
}

/// Return only non-sensitive paths.
pub fn filter_sensitive_paths<'a>(paths: &'a [&'a Path]) -> Vec<&'a Path> {
    paths
        .iter()
        .filter(|p| !is_sensitive_file(p))
        .copied()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sensitive_files() {
        assert!(is_sensitive_file(Path::new(".env")));
        assert!(is_sensitive_file(Path::new(".env.local")));
        assert!(is_sensitive_file(Path::new("server.key")));
        assert!(is_sensitive_file(Path::new("cert.pem")));
        assert!(is_sensitive_file(Path::new("id_rsa")));
        assert!(is_sensitive_file(Path::new("credentials.json")));
    }

    #[test]
    fn test_safe_files() {
        assert!(!is_sensitive_file(Path::new("main.py")));
        assert!(!is_sensitive_file(Path::new("README.md")));
        assert!(!is_sensitive_file(Path::new("config.toml")));
    }

    #[test]
    fn test_sensitive_symlink_alias() {
        let dir = tempfile::tempdir().unwrap();
        let sensitive = dir.path().join(".env");
        std::fs::write(&sensitive, "SENSITIVE-SENTINEL").unwrap();
        let alias = dir.path().join("notes.txt");

        #[cfg(unix)]
        let link_result = std::os::unix::fs::symlink(&sensitive, &alias);
        #[cfg(windows)]
        let link_result = std::os::windows::fs::symlink_file(&sensitive, &alias);
        if link_result.is_err() {
            return;
        }

        assert!(is_sensitive_file(&alias));
    }

    #[test]
    fn test_sensitive_symlink_alias_to_missing_target() {
        let dir = tempfile::tempdir().unwrap();
        let alias = dir.path().join("notes.txt");
        let sensitive = dir.path().join(".env");

        #[cfg(unix)]
        let link_result = std::os::unix::fs::symlink(&sensitive, &alias);
        #[cfg(windows)]
        let link_result = std::os::windows::fs::symlink_file(&sensitive, &alias);
        if link_result.is_err() {
            return;
        }

        assert!(is_sensitive_file(&alias));
    }
}
